### ISSUE TITLE
Start and stop at non-zero

### ISSUE THREAD

**Issue by bthj** [ircam-ismm/node-web-audio-api#142]
Hi, I was wondering if it's a feature of node-web-audio-api / web-audio-api-rs to ensure sounds start and stop at zero sample values, and if there is some ramping towards those values?

This would be good to avoid clicks, but now I'm looking into the possibility of rendering a single sound as several fractions, e.g. a 2s sound as four 0.5s parts, to obtain initial audio buffer results faster and approach realtime-ish feedback: I'm using the OfflineAudioContext where signals come through data arrays passed in with setValueCurveAtTime. I can compute those data arrays in smaller segments and then pass the values through the DSP graph enabled by the Web Audio API. When trying this I observe amplitude dips at the partition boundaries. I've tried to ensure this is not happening in my computation (could still be), so I'm wondering if this is a behaviour of the Web Audio API implementation, and if so, whether I could optionally disable it? 

Another option would be to render a slight excess of each partition, for overlapping, but then I also wonder if there is a known duration in samples for fading to and from zero?

**Comment by b-ma** [comment]
Hey, 

There is no such thing as ensuring that the sound from an `AudioBufferSourceNode` starts and stops at zero, it just plays the samples that the buffer contains. If you want to have such a ramp you need to pipe the source into a `GainNode` and schedule the envelop automation manually.

Not completely sure I understand well your use case but if I do, you should be able to stitch subsequent audio buffer together without further processing now (cross-fading or something). This is something we fixed [recently](https://github.com/orottier/web-audio-api-rs/blob/main/CHANGELOG.md#version-110-2024-12-11) and that has been shipped with v0.21.4. Then with just proper scheduling of `start()`, something like this should be ok now:

```js
let now = audioContext.currentTime;

const src1 = audioContext.createBufferSource();
src1.buffer = buffer1;
src1.start(now);
// update now for next buffer
now += buffer1.duration;

const src2 = audioContext.createBufferSource();
src2.buffer = buffer2;
src2.start(now); // start exactly at the end of buffer 1
```

Let me know!

**Comment by bthj** [comment]
Thanks for confirming @b-ma 

I abstracted what I'm trying, away from the rest of [my code](https://github.com/synth-is/kromosynth), in the following snippet, where audio data is split into segments, and each segment is rendered via the Web Audio API implementation (where the rendering should start immediately, so I'm not using the `.start` attribute). Then the rendered segments are concatenated (as a potential receiver might do) and the result is seamless, without the unwanted dips / ramps I've observed, so that must be introduced somewhere else - will have a better look into that.

In any case this may not be a good approach: it could be better to produce the channel data segments in AudioWorklets; will also look into that.

```
import { OfflineAudioContext } from 'node-web-audio-api';
import toWav from 'audiobuffer-to-wav';
import fs from 'fs';

const SAMPLE_RATE = 48000;

const duration = 10;
const frequency = 440;
const samples = duration * SAMPLE_RATE;
const sineWave = new Float32Array(samples);

for (let i = 0; i < samples; i++) {
  sineWave[i] = Math.sin(2 * Math.PI * frequency * i / SAMPLE_RATE);
}

const segmentLength = samples / 10;
const segments = [];

for (let i = 0; i < 10; i++) {
  segments.push(sineWave.subarray(i * segmentLength, (i + 1) * segmentLength));
}

const renderedSegments = [];

// render each segment individually
for (const segment of segments) {
  let offline = new OfflineAudioContext(1, segment.length, SAMPLE_RATE);
  const segmentBuffer = offline.createBuffer(1, segment.length, SAMPLE_RATE);
  segmentBuffer.copyToChannel(segment, 0);

  const segmentSource = offline.createBufferSource();
  segmentSource.buffer = segmentBuffer;
  segmentSource.connect(offline.destination);
  segmentSource.start();

  const renderedBuffer = await offline.startRendering();
  renderedSegments.push(renderedBuffer.getChannelData(0));
}

// concatenate the rendered segments
const concatenatedBuffer = new Float32Array(renderedSegments.reduce((acc, val) => acc + val.length, 0));
let offset = 0;
for (const segment of renderedSegments) {
  concatenatedBuffer.set(segment, offset);
  offset += segment.length;
}

let offline = new OfflineAudioContext(1, 10*SAMPLE_RATE, SAMPLE_RATE);
const concatenatedAudioBuffer = offline.createBuffer(1, concatenatedBuffer.length, SAMPLE_RATE);
concatenatedAudioBuffer.copyToChannel(concatenatedBuffer, 0);

const concatenatedAudioBufferWav = toWav(concatenatedAudioBuffer);
const concatenatedAudioBufferWavBuffer = Buffer.from(new Uint8Array(concatenatedAudioBufferWav));
fs.writeFileSync("concatenated_oscillator.wav", concatenatedAudioBufferWavBuffer);
```
